engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java

```diff
@@ -135,13 +146,103 @@ private Either<List<DirectBuffer>, CatchEventTuple> findCatchEventInScope(
   }
 
   public Optional<CatchEventTuple> findEscalationCatchEvent(
-      final DirectBuffer escalationCode, final ElementInstance instance) {
-    // TODO walk through the scope hierarchy and look for a matching catch event
+      final DirectBuffer escalationCode, ElementInstance instance) {
+    // walk through the scope hierarchy and look for a matching catch event
+    while (instance != null && instance.isActive()) {
+      final var instanceRecord = instance.getValue();
+      final var process = getProcess(instanceRecord.getProcessDefinitionKey());
+
+      final var found = findEscalationCatchEventInProcess(escalationCode, process, instance);
+      if (found.isPresent()) {
+        return found;
+      }
+
+      // find in parent process instance if exists
+      final var parentElementInstanceKey = instanceRecord.getParentElementInstanceKey();
+      instance = elementInstanceState.getInstance(parentElementInstanceKey);
```

Comment:
🙃 Please note that this isn't a must. This code is fine and readable. Personally I would've used recursion for this:

```java
      return findEscalationCatchEventInProcess(escalationCode, process, instance)
          .or(
              () -> {
                final ElementInstance parentElementInstanceKey =
                    elementInstanceState.getInstance(instanceRecord.getParentElementInstanceKey());
                return findEscalationCatchEvent(escalationCode, parentElementInstanceKey);
              });
```

This would also involve changing from a `while`-loop to an `if`-statement